### ISSUE TITLE
Include `jackson-datatype-hibernate5` in `quarkus-resteasy-reactive-jackson` and `quarkus-resteasy-jackson`

### ISSUE THREAD

**Issue by agreedSkiing** [quarkusio/quarkus#30096]
### Description

Since Quarkus 2.X is still based on Hibernate Orm 5.X why not include `jackson-datatype-hibernate5` from the https://github.com/FasterXML/jackson-datatype-hibernate libraries?

This would reduce the amount LazyInitializationException that can happen when creating test projects or creating audit logs of changes in a database.

**Comment by geoand** [comment]
I'd say that we should probably not do something that would make migration to Hibernate 6 more difficult

**Comment by agreedSkiing** [comment]
Right, jackson-datatype-hibernate is not fully compatible with Hibernate 6 😟 and just looked through their merge request that was merged for hibernate 6 and there seems to be some problems with jdk-11 and older test so they won't release jackson-datatype-hibernate6 yet.

https://github.com/FasterXML/jackson-datatype-hibernate/pull/162#issuecomment-1358603111

**Comment by agreedSkiing** [comment]
Oh, wait a second, the last merge fixed all the test errors 😮, so now they only need to tag and release a version to maven central. Question is when will that happen? (who knows) 

https://github.com/FasterXML/jackson-datatype-hibernate/pull/163

**Comment by yrodiere** [comment]
> I'm a bit skeptical about how this works, especially in terms of transaction boundaries and the relaxed approch of representing unloaded relations as null; is this really handy? It _seems_ to me people would be better off by mapping explicit result objects.

Agreed. Loading data outside of transaction boundaries is dangerous, personally it'll be DTO any day.

So I don't personally want to encourage people to use [jackson-datatype-hibernate](https://github.com/FasterXML/jackson-datatype-hibernate), and I'd be tempted to just close this.

I'm wondering, however, if we should propose an easy-to-use, safe-ish alternative: we could make sure that serialization of entities returned from transactional methods actually happens within transaction boundaries, when the session/connection is still available.
It's still dodgy, because, well, serializing an entity graph is no easy task and can result in serializing the whole database; but I can see it being useful, and it would avoid the most common pitfalls. And then, maybe `jackson-datatype-hibernate` could be helpful without also being dangerous.

**Comment by geoand** [comment]
@yrodiere that is a very interesting idea. In Quarkus REST we could probably introduce some new annotation (or even repurpose the existing `@Transactional` one, although this is debatable) to ensure that the transaction is closed **after** the returned entity has been written to the body of the response. 
This could potentially create other problems for us, but it's probably worth experimenting with.

**Comment by yrodiere** [comment]
> to ensure that the transaction is closed after the returned entity has been written to the body of the response.

I'm not sure how serialization is done exactly, but if returned objects are first serialized into an in-memory buffer, that would be when I would close the transaction -- not when we finished writing entities. Reason is we want serialization errors to impact the HTTP status, and that one is returned first, before we start writing the response body.

If there is no buffering by default, I think this would be a reasonable trade-off to enable it in this case: if you want lazy loading during serialization, we need buffering 🤷 

> we could probably introduce some new annotation (or even repurpose the existing @Transactional one, although this is debatable)

I suspect people wanting this will want it almost everywhere, so a new annotation would be quite verbose...

I'm also not sure why someone would *not* want it -- at least if done as suggested above... Performance, maybe -- but then we could possibly disable this feature for methods whose return type doesn't involve entity types at any point (including in nested data). That detection could double as a UX improvement by issuing warnings when users attempt to serialize entity types but don't use `@Transactional` on their endpoint -- though there might be legitimate cases, so this needs more thought and/or a way to disable warnings.

Maybe a first attempt could provide a global flag, `quarkus.rest.serialize-in-transaction` or something, to opt out if there's a problem, but otherwise enable this all automatically as soon as methods are annotated with `@Transaction` (and have a relevant return type, i.e. one involving entities).

**Comment by geoand** [comment]
> I'm not sure how serialization is done exactly, but if returned objects are first serialized into an in-memory buffer, that would be when I would close the transaction

That can't really be done because we don't control how the serialization library does its work. We can only react to serialization having completed (successfully or not).

The reason I am thinking of a new annotation is simply to make users aware that the transaction will now close not when REST method itself  completes, but when the data has actually been written out.

**Comment by Sanne** [comment]
> [@yrodiere](https://github.com/yrodiere) that is a very interesting idea. In Quarkus REST we could probably introduce some new annotation (or even repurpose the existing `@Transactional` one, although this is debatable) to ensure that the transaction is closed **after** the returned entity has been written to the body of the response. This could potentially create other problems for us, but it's probably worth experimenting with.

A word of caution: unless we know for sure that a transaction is strictly read-only, there are possibilities that the transaction might fail. Foreign key violations, duplicate entries, unacceptable data ranges, any state checking triggers...

It's generally not a good idea to keep the transaction running while rendering a response, because if there is an error one might want to redirect or show different content, or sometimes even retry the operation (for example whe using optimistic locking). How are you going to do any of these if you're already mid-way rendered the response?

The Seam framework had some interesting patterns but didn't scale very well. I don't have a silver-bullet solution, just raising awareness that this needs to be somewhat thought through, and hopefully do some research on prior art - especially failed approaches people already faced.

**Comment by geoand** [comment]
Thanks for bringing that up @Sanne !

**Comment by yrodiere** [comment]
> > I'm not sure how serialization is done exactly, but if returned objects are first serialized into an in-memory buffer, that would be when I would close the transaction
> 
> That can't really be done because we don't control how the serialization library does its work. We can only react to serialization having completed (successfully or not).

Hm. If we can't end the transaction before we return the status, then the transaction commit may fail and we won't even tell the user. 

We would need to do serialization in a buffer, then commit then transaction, then determine the response status... otherwise it's not even worth talking about.

**Comment by geoand** [comment]
> We would need to do serialization in a buffer, then commit the transaction, then determine the response status... otherwise it's not even worth talking about.

Understood, so let's stop talking about it in that case :)

**Comment by yrodiere** [comment]
@geoand last-ditch attempt: can we take serialization into our own hands, and somehow make quarkus-rest "believe" the method returns a `byte[]`/`String`? So that serialization happens earlier, and we get a chance to perform a commit?

**Comment by geoand** [comment]
Under the right circumstances, yeah, that is likely doable.

The circumstances I can think of off the top of my head are:

* No async types are being used
* No custom `MessageBodyWriter` types are in play. 
* We can figure out the proper `MessageBodyWriter` at build time (I am not 100% on this one, but it's likely necessary)

**Comment by yrodiere** [comment]
Understood. Created #44918 to keep track of this idea.

In the meantime, I would consider this issue (#30096) blocked, because it's only reasonable if we implement #44918 first IMO.

**Comment by yrodiere** [comment]
> In the meantime, I would consider this issue ([#30096](https://github.com/quarkusio/quarkus/issues/30096)) blocked, because it's only reasonable if we implement [#44918](https://github.com/quarkusio/quarkus/issues/44918) first IMO.

Discussions on #44918 proved that this was not a good idea.

We should really look into `jackson-datatype-hibernate*`; it's way past time.

**Comment by FroMage** [comment]
Well, I don't know… how many times are we going to have to point users away from serialising entities using Jackson?

Wouldn't this only make it work _that little bit more_ until entity relations break it, or worse, security leaks happen?

Unless these two issues are solved by `jackson-datatype-hibernate` (they may be, I haven't been able to find their documentation from their github project) then I'm not sure it's worth discussing it, and we should focus on a proper DTO story, no?

I have noticed that `@JsonIdentityInfo` annotations would solve the issue of any realistic entity graph, as well as `@JsonManagedReference/@JsonBackReference`, so OK, if we start adding lots of annotations to the JPA model, we _may_ not get an infinite recursion error on serialising. But remains the security risks. I truely feel it's irresponsible of us to suggest using entities as data objects.

**Comment by agreedSkiing** [comment]
Been wondering, also throught that this feature/enhancement was closed, is there no way to implement a native quarkus check that blocks this kind of "bad" use of entities?

I currently feel like it would be a better approach to present users with a hard error "No this is not allowed with Quarkus, implement DTOs instead" and link them to Quarkus goos practices for REST.

**Comment by yrodiere** [comment]
> I truely feel it's irresponsible of us to suggest using entities as data objects.

I encourage you to talk to Gavin directly.

My position right now is almost exactly between you two.

**Comment by FroMage** [comment]
Wait, Gavin is in favour of using entities for REST exchanges? That'd be news to me.

**Comment by yrodiere** [comment]
> Wait, Gavin is in favour of using entities for REST exchanges? That'd be news to me.

In favor of making sure it works properly, at least. Like I said, please talk to him to find out more :)

EDIT: BTW he does have good arguments, like the fact there is everything available in Hibernate SPIs to make it work to some extent (e.g. preventing lazy-loading after session is closed, omitting non-loaded values instead of throwing exceptions).

**Comment by FroMage** [comment]
I do have to talk to him indeed, thanks.

**Comment by FroMage** [comment]
I've talked to @gavinking and here are my notes:

The current situation is not great. Out of the box, we cannot use realistic entities as DTOs because they break down for relations, and they open a huge security risk for any realistic entity (by exposing internal fields and worse: allowing users to upload entities with internal fields and using `merge` to inject unwanted data).

We never invested the time to properly implement or support any DTO framework, instead we've sent a confusing message because in demos we always use entities as DTOs, while telling users (if we remember) that they should not. And I've felt safe knowing they could not use real world entities as DTOs knowing they would break at any real model before security became an issue.

We do point users towards [Json Views](https://quarkus.io/guides/rest#jsonview-support) which is one way of solving both the relation cycle and the security issues by using entities as a DTO by defining an opt-in set of fields to expose.

We also have the [`@SecureField`](https://quarkus.io/guides/rest#secure-serialization) annotation which solves one aspect of security (for serialisation only, I assume, as the docs do not mention deserialisation) by restricting the presence of fields based on the current user roles, when using entities as a DTO.

And we have a [Quarkiverse extension for MapStruct](https://docs.quarkiverse.io/quarkus-mapstruct/dev/index.html) which allows users to define their own separate DTO classes and generates translator methods to convert an entity to its DTO both ways, using an annotation processor, making this type-safe, but requiring a parallel class hierarchy to represent DTOs.

By not picking a side, we've led users to figure things out by themselves and provided an OOTB experience that is not great.

The `jackson-datatype-hibernate` Jackson Module allows us to configure several things:

- What happens when Jackson meets an unloaded relation (by default, it serialises it as `null`, but it could force lazy-loading or serialise an ID)
- Respect the `@Transient` annotation to ignore the property (defaults to true)
- Other aspects I thought were less important

Note that it does not extend the transaction boundary, so the lazy-loading forcing aspect is pretty much moot, because by the time the serialiser runs, the transaction will be over and so the loading will fail.

We could improve the experience for people wanting to use entities as DTOs out of the box, by:

- Having a central guide to explain how to map database entities to JSON, what the options are, and what the security implications are
- Tweaking either entities or the default Jackson module to ignore all properties of entities by default. Users would have to mark exposed fields using `@JsonProperty`. This would solve the security issues by requiring users to opt-in.
- If an entity is annotated with `@JsonExposeAllProperties` (or an existing equivalent Jackson annotation) then all properties would be exposed. This would have a strong warning, and be mostly useful for demos.
- Automatically adding `@JsonManagedReference/@JsonBackReference` on relations (not entirely sure about this one, we could let users figure this one out, but they'd start their experience with a cycle exception which is not great)
- Making sure unloaded relations are never loaded by Jackson. If users want them to be serialised, they can include them in the fetch plan to make sure they are loaded. This solves the transaction-boundary exception and forces users to be explicit.
- Figure out and explain how `@JsonIdentityInfo` should be used for relations
- I guess figure out how all of this impacts/improves/fixes the `panache-rest-data-json` module

Between this, and Json Views and `@SecureField` we'd have a pretty strong story for using Jackson and entities as DTOs.

Now, I don't think this requires `jackson-datatype-hibernate`, because it doesn't have all the features I just described, and I don't think I agree that (jpa's) `@Transient` should imply `@JsonIgnore` because I'm not sure how else to expose a JSON-only property that is not a JPA property. It might be better to implement our own Jackson module for Hibernate for Quarkus.

As for the `MapStruct` story, it definitely is an alternate take on the subject, and we could document it properly, and we could even improve it by making its annotation processor (or the hibernate one, perhaps) automatically generate DTOs for existing entities, which MapStruct can then provide translator methods for. 

For example:

```java
@Entity
public class MyEntity {
 @Id
 @AutoGenerated
 public Long id;
 @DTOField
 publid String name;
 public String other;
}
```

Would automatically generate:

```java
@Generated
public record MyEntityDTO(String name) {

 public static MyEntityDTO fromEntity(MyEntity entity) {...}
 public static MyEntity toEntity(MyEntityDTO entity) {...}
 // or
 public MyEntity toEntity() {…}
}
```

All of this suggests I should open a new issue to discuss this, rather than implement this issue about "just" including `jackson-datatype-hibernate`.

**Comment by geoand** [comment]
> I should open a new issue to discuss this

+1

**Comment by yrodiere** [comment]
+1 for a separate issue if you want to discuss going beyond `jackson-datatype-hibernate`.

BTW these are all great ideas but I worry about the cost (both initial and of maintenance) of reinventing the wheel. Reusing what's out there might be more realistic if we have other problems to tackle in parallel...

**Comment by FroMage** [comment]
Well, except I do feel that `jackson-datatype-hibernate` is a bad idea because it provides behaviour that is not what we want, and does not provide behaviour that our users need.

**Comment by agreedSkiing** [comment]
Based on the lengthy discussions that has been has, is it not time to close this request/issue then? I feel like that there is a consensus which can be summarized to (from my brain):

Don't use entities as DTO and Quarkus (teamm) does not recommend using entities as DTO since they might leak information or cause security risk when persisting data.

Instead this discussion will be branched out to other issues/requests for discussions and or solutions in the long run?

**Comment by FroMage** [comment]
OK, let's move it to https://github.com/quarkusio/quarkus/issues/53263 then
